src/ts/features/block-navigation-mode/blockNavigation.ts

```diff
@@ -0,0 +1,177 @@
+import {assumeExists} from '../../utils/assert'
+import {Selectors} from '../../roam/roam-selectors'
+
+export type Panel = 'MAIN' | 'SIDEBAR'
+
+type State = {
+    mainBlockId: string | null
+    sideBlockId: string | null
+    panel: Panel
+}
+
+export const state: State = {
+    mainBlockId: null,
+    sideBlockId: null,
+    panel: 'MAIN',
+}
+
+export const getFocusedPanel = (): HTMLElement => {
+    if (state.panel === 'MAIN') {
+        const articleElement = assumeExists(document.querySelector(Selectors.mainContent))
+        return assumeExists(articleElement.parentElement)
+    }
+    if (state.panel === 'SIDEBAR') {
+        return assumeExists(document.querySelector(Selectors.sidebarContent))
+    }
+    throw new Error(`Unexpected panel ${state.panel}`)
+}
+
+const getPanel = (panel?: Panel): HTMLElement | null => {
+    if (!panel) {
+        return getFocusedPanel()
+    }
+    return panel === 'MAIN' ? mainPanel() : rightPanel()
+}
+
+const rightPanel = (): HTMLElement | null => document.querySelector(Selectors.sidebarContent)
+
+const mainPanel = (): HTMLElement | null => document.querySelector(Selectors.mainContent)?.parentElement || null
+
+export const ensureMainPanelHasBlockSelected = () => {
+    if (!getBlock(state.mainBlockId)) {
+        const mainPanel = assumeExists(getPanel('MAIN'))
+        state.mainBlockId = mainPanel.querySelector(Selectors.block)?.id || null
+    }
+}
+
+export const ensureRightPanelHasBlockSelected = () => {
+    const rightPanel = getPanel('SIDEBAR')
+    if (rightPanel && !getBlock(state.sideBlockId)) {
+        state.sideBlockId = rightPanel.querySelector(Selectors.block)?.id || null
+    }
+}
+
+export const ensureFocusedPanelHasBlockSelected = () => {
+    if (state.panel === 'MAIN') {
+        ensureMainPanelHasBlockSelected()
+    } else {
+        ensureRightPanelHasBlockSelected()
+    }
+}
+
+const blockIdRelative = (block: HTMLElement, blocksToJump: number, panel: Element): string => {
+    const blocks = panel.querySelectorAll(Selectors.block)
+
+    let blockIndex
+    // @ts-ignore NodeList.entries() does exist in Chrome/FF:
+    // https://developer.mozilla.org/en-US/docs/Web/API/NodeList/entries
+    for (let [i, _block] of blocks.entries()) {
```

Comment:
wouldn't using `forEach` which ts recognizes give us the same result?

Reply:
I wanted to short circuit if I found the block early, but I guess I'd have to loop through every block if the last one was selected anyways.

I'll switch to `forEach`